import_obfuscation.go

```diff
@@ -0,0 +1,572 @@
+		}
+	}
+
+	var sb strings.Builder
+	var buf bytes.Buffer
+	garbledImports := make(map[string]string)
+	for _, p := range privatePkgs {
+		// log.Printf("++ Obfuscating object file for %s ++", p.pkg.ImportPath)
+		for _, am := range p.pkg.ArchiveMembers {
+			// log.Printf("\t## Obfuscating archive member %s ##", am.ArchiveHeader.Name)
+
+			// skip objects that are not used by the linker, or that do not contain
+			// any Go symbol info
+			if am.IsCompilerObj() || am.IsDataObj() {
+				continue
+			}
+
+			// add all private import paths to a list to garble
+			var privImports privateImports
+			privImports.privatePaths, privImports.privateNames = explodeImportPath(p.pkg.ImportPath)
+			// the main package might not have the import path "main" due to modules,
+			// so add "main" to private import paths
+			if p.pkg.ImportPath == buildInfo.firstImport {
+				privImports.privatePaths = append(privImports.privatePaths, "main")
+			}
+
+			initImport := func(imp string) string {
+				if !isPrivate(imp) {
+					return imp
+				}
+
+				privPaths, privNames := explodeImportPath(imp)
+				privImports.privatePaths = append(privImports.privatePaths, privPaths...)
+				privImports.privateNames = append(privImports.privateNames, privNames...)
+
+				return hashImport(imp, garbledImports)
+			}
+
+			for i := range am.Imports {
+				am.Imports[i].Pkg = initImport(am.Imports[i].Pkg)
+			}
+			for i := range am.Packages {
+				am.Packages[i] = initImport(am.Packages[i])
+			}
+
+			// move imports that contain another import as a substring to the front,
+			// so that the shorter import will not match first and leak part of an
+			// import path
+			sort.Slice(privImports.privatePaths, func(i, j int) bool {
+				iSlashes := strings.Count(privImports.privatePaths[i], "/")
+				jSlashes := strings.Count(privImports.privatePaths[j], "/")
+				// sort by number of slashes first, then alphabetically
+				if iSlashes == jSlashes {
+					return privImports.privatePaths[i] > privImports.privatePaths[j]
+				}
+				return iSlashes > jSlashes
+			})
+			sort.Slice(privImports.privateNames, func(i, j int) bool {
+				return privImports.privateNames[i] > privImports.privateNames[j]
+			})
+			privImports.privatePaths = dedupStrings(privImports.privatePaths)
+			privImports.privateNames = dedupStrings(privImports.privateNames)
+
+			// no private import paths, nothing to garble
+			if len(privImports.privatePaths) == 0 {
+				continue
+			}
+			// log.Printf("\t== Private imports: %v ==\n", privImports)
+
+			// garble all private import paths in all symbol names
+			lists := [][]*goobj2.Sym{am.SymDefs, am.NonPkgSymDefs, am.NonPkgSymRefs}
+			for _, list := range lists {
+				for _, s := range list {
+					// skip debug symbols, and remove the debug symbol's data to save space
+					if s.Kind >= goobj2.SDWARFINFO && s.Kind <= goobj2.SDWARFLINES {
+						s.Data = nil
+						continue
+					}
+
+					// skip local asm symbols. For some reason garbling these breaks things
```

Comment:
What are local asm symbols?

Reply:
ASM symbols defined in the 'main' package